launcher/JavaDownloader.cpp

```diff
@@ -0,0 +1,321 @@
+    connect(netJob, &NetJob::failed, this, &JavaDownloader::emitFailed);
+
+    connect(this, &Task::aborted, netJob, &NetJob::abort);
+
+    connect(netJob, &NetJob::succeeded, [response, OS, isLegacy, this, netJob] {
+        QJsonParseError parse_error{};
+        QJsonDocument doc = QJsonDocument::fromJson(*response, &parse_error);
+        if (parse_error.error != QJsonParseError::NoError) {
+            qWarning() << "Error while parsing JSON response at " << parse_error.offset << " reason: " << parse_error.errorString();
+            qWarning() << *response;
+            return;
+        }
+        auto versionArray = Json::ensureArray(Json::ensureObject(doc.object(), OS), isLegacy ? "jre-legacy" : "java-runtime-gamma");
+        if (!versionArray.empty()) {
+            parseMojangManifest(isLegacy, versionArray);
+
+        } else {
+            // mojang does not have a JRE for us, let's get azul zulu
+            downloadAzulMeta(OS, isLegacy, netJob);
+        }
+    });
+
+    netJob->start();
+}
+void JavaDownloader::parseMojangManifest(bool isLegacy, const QJsonArray& versionArray)
+{
+    setStatus(tr("Downloading java from Mojang"));
+    auto url = Json::ensureString(Json::ensureObject(Json::ensureObject(versionArray[0]), "manifest"), "url");
+    auto download = new NetJob(QString("JRE::DownloadJava"), APPLICATION->network());
+    auto files = new QByteArray();
+
+    download->addNetAction(Net::Download::makeByteArray(QUrl(url), files));
+
+    connect(download, &NetJob::finished, [download, files, this] {
+        disconnect(this, &Task::aborted, download, &NetJob::abort);
+        download->deleteLater();
+        delete files;
+    });
+    connect(download, &NetJob::progress, this, &JavaDownloader::progress);
+    connect(download, &NetJob::failed, this, &JavaDownloader::emitFailed);
+    connect(this, &Task::aborted, download, &NetJob::abort);
+
+    connect(download, &NetJob::succeeded, [files, isLegacy, this] {
+        QJsonParseError parse_error{};
+        QJsonDocument doc = QJsonDocument::fromJson(*files, &parse_error);
+        if (parse_error.error != QJsonParseError::NoError) {
+            qWarning() << "Error while parsing JSON response at " << parse_error.offset << " reason: " << parse_error.errorString();
+            qWarning() << *files;
+            return;
+        }
+        downloadMojangJava(isLegacy, doc);
+    });
+    download->start();
+}
+void JavaDownloader::downloadMojangJava(bool isLegacy, const QJsonDocument& doc)
+{  // valid json doc, begin making jre spot
+    auto output = FS::PathCombine(QString("java"), (isLegacy ? "java-legacy" : "java-current"));
+    FS::ensureFolderPathExists(output);
+    std::vector<File> toDownload;
+    auto list = Json::ensureObject(Json::ensureObject(doc.object()), "files");
+    for (const auto& paths : list.keys()) {
+        auto file = FS::PathCombine(output, paths);
+
+        auto type = Json::requireString(Json::requireObject(list, paths), "type");
```

Comment:
Those requires will probably cause a crash on some network or 3rd-party issues, i don't think it's a good idea to require fields from 3rd-party requests without a try-catch enclosure :grimacing: